# Artificial Neural Network

### Importing the libraries

In [1]:
import pandas as pd
import numpy as np
import tensorflow as tf

In [2]:
tf.__version__

'2.3.0'

## Part 1 - Data Preprocessing

### Importing the dataset

In [3]:
dataset = pd.read_csv("500_Person_Gender_Height_Weight_Index.csv")
X = dataset.iloc[:,:-1].values
y = dataset.iloc[:,-1].values

### Encoding categorical data

Label Encoding the "Gender" column

In [4]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
X[:, 0] = le.fit_transform(X[:, 0])

### Splitting the dataset into the Training set and Test set

In [5]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 0)

### Feature Scaling

In [6]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

## Part 2 - Building the ANN

### Initializing the ANN

In [7]:
ann = tf.keras.models.Sequential()

### Adding the input layer and the first hidden layer

In [8]:
ann.add(tf.keras.layers.Dense(units=9, activation="softmax"))

### Adding the second hidden layer

In [9]:
ann.add(tf.keras.layers.Dense(units=9, activation="softmax"))

### Adding the output layer

In [10]:
ann.add(tf.keras.layers.Dense(units=6, activation='softmax'))

## Part 3 - Training the ANN

### Compiling the ANN

In [11]:
ann.compile(optimizer= "adam", loss= "sparse_categorical_crossentropy" , metrics= ["accuracy"])

### Training the ANN on the Training set

In [12]:
ann.fit(X_train, y_train, batch_size= 32, epochs= 600)

Epoch 1/600
12/12 [==============================] - 0s 1ms/step - loss: 1.7771 - accuracy: 0.1333
Epoch 2/600
12/12 [==============================] - 0s 1ms/step - loss: 1.7629 - accuracy: 0.1333
Epoch 3/600
12/12 [==============================] - 0s 1ms/step - loss: 1.7490 - accuracy: 0.1333
Epoch 4/600
12/12 [==============================] - 0s 1ms/step - loss: 1.7361 - accuracy: 0.1440
Epoch 5/600
12/12 [==============================] - 0s 2ms/step - loss: 1.7237 - accuracy: 0.3813
Epoch 6/600
12/12 [==============================] - 0s 1ms/step - loss: 1.7115 - accuracy: 0.3947
Epoch 7/600
12/12 [==============================] - 0s 1ms/step - loss: 1.7003 - accuracy: 0.3947
Epoch 8/600
12/12 [==============================] - 0s 1ms/step - loss: 1.6894 - accuracy: 0.3947
Epoch 9/600
12/12 [==============================] - 0s 1ms/step - loss: 1.6791 - accuracy: 0.3947
Epoch 10/600
12/12 [==============================] - 0s 1ms/step - loss: 1.6693 - accuracy: 0.3947
Epoch 11/

## Part 4 - Making the predictions and evaluating the model

### Predicting the Test set results

In [16]:
y_pred = ann.predict(X_test)
y_pred = (y_pred > 0.5)

In [14]:
y_pred_index = []
for i in y_pred:
  index = np.where(i == True)
  if len(index[0]) == 0:
    y_pred_index.append(0)
  else:
    y_pred_index.append(int(index[0]))
print(y_pred_index)

[5, 5, 4, 5, 5, 5, 3, 5, 4, 5, 2, 4, 2, 5, 5, 4, 5, 4, 5, 4, 4, 3, 0, 4, 2, 2, 5, 4, 4, 0, 4, 5, 5, 5, 4, 2, 5, 4, 5, 2, 5, 4, 4, 5, 4, 2, 5, 0, 5, 5, 4, 4, 0, 5, 5, 3, 4, 4, 2, 4, 4, 4, 4, 2, 4, 4, 2, 3, 5, 4, 5, 5, 4, 4, 4, 5, 4, 5, 5, 5, 0, 3, 4, 5, 5, 2, 5, 5, 5, 4, 3, 2, 3, 2, 3, 2, 4, 4, 2, 0, 5, 5, 4, 3, 5, 5, 2, 5, 4, 4, 4, 5, 3, 4, 5, 5, 5, 4, 4, 5, 0, 5, 5, 5, 5]


### Making the Confusion Matrix

In [15]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_pred_index, y_test)
print("CM :", cm )
accuracy_score(y_pred_index, y_test)

CM : [[ 0  0  2  4  0  1]
 [ 0  0  0  0  0  0]
 [ 3  4  9  0  0  0]
 [ 0  0  1  9  0  0]
 [ 0  0  0  5 37  0]
 [ 0  0  0  0  1 49]]


0.832

# Conclusion:

> To evaluate that BMI index I have used data of 500 person Gender, Height, Weight and Index and found that artificial neural network get the accuracy of 83.20 % whereas Random Forest regressor got the accuracy of 95 %, So it is too much important to select the right algorithm for the respective problem